jetty-core/jetty-http2/jetty-http2-common/src/main/java/org/eclipse/jetty/http2/HTTP2Session.java

```diff
@@ -2237,19 +2259,76 @@ private MetaData.Request extractMetaDataRequest(StreamFrame frame)
 
         private int reserveSlot(Slot slot, int streamId, Consumer<Throwable> fail)
         {
+            if (streamId < 0 || (streamId > 0 && !isLocalStream(streamId)))
+            {
+                fail.accept(new IllegalArgumentException("invalid stream id " + streamId));
+                return 0;
+            }
+
+            boolean created = false;
+            int reservedStreamId = 0;
             Throwable failure = null;
-            boolean reserved = false;
             try (AutoLock ignored = lock.lock())
             {
                 // SPEC: cannot create new streams after receiving a GOAWAY.
                 if (closed == CloseState.NOT_CLOSED)
                 {
-                    if (streamId <= 0)
+                    if (streamId == 0)
                     {
-                        streamId = localStreamIds.getAndAdd(2);
-                        reserved = true;
+                        // Stream id generated internally.
+                        reservedStreamId = localStreamIds.getAndAdd(2);
+                        // Check for overflow.
+                        if (reservedStreamId > 0)
```

Comment:
This test is surprising to me.  I would have expected a test against a value derived from the maxUsage method, rather than just knowing that the max Id is at arithmetic overflow.